lib/routes/key_data.js

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const Joi = require('joi');
+
+const AppError = require('../error');
+const db = require('../db');
+const logger = require('../logging')('routes.key_data');
+const P = require('../promise');
+const validators = require('../validators');
+const verify = require('../browserid');
+const Scope = require('../scope');
+
+const ADDITIONAL_KEY_MATERIAL = Buffer.alloc(32).toString('hex');
+
+module.exports = {
+  validate: {
+    params: {
+      client_id: validators.clientId
+    },
+    payload: {
+      assertion: validators.assertion.required(),
+      scope: Joi.string()
+    }
+  },
+  response: {
+    schema: Joi.object().pattern(/^/, [
+      Joi.object({
+        identifier: Joi.string().required(),
+        keyMaterial: Joi.string().required(),
+        timestamp: Joi.number().required()
+      })
+    ])
+  },
+  handler: function keyDataRoute(req, reply) {
+    logger.debug('keyDataRoute.start', {
+      params: req.params,
+      payload: req.payload
+    });
+
+    const requestedScopes = new Set(Scope(req.payload.scope).values());
+    const requestedClientId = req.params.client_id;
+
+    P.all([
+      verify(req.payload.assertion),
+      db.getClient(Buffer.from(requestedClientId, 'hex')).then((client) => {
+        if (client) {
+          const allowedScopes = new Set(Scope(client.allowedScopes || []).values());
+          // find all requested scopes in allowed scopes
+          const scopes = new Set([...allowedScopes].filter((x) => requestedScopes.has(x)));
+          const scopeRequests = [];
```

Comment:
There's a lot of casting to `Set` here that doesn't seem necessary.  Could this be done more simply with something like the following?

```
const requestedScopes = Scope(req.payload.scope)
...
if (client) {
  const scopeRequests = [];
  const allowedScopes = Scope(client.allowedScopes)
  requestedScopes.values().forEach((s) => {
    if (allowedScopes.has(s)) {
      scopeRequests.push(db.getScope(s))
    }
  })
}
```

We could consider adding more helpers to the `Scope` class to make this easier, e.g. `intersect` and/or `filter` and/or ` forEach`.  But that can be for another day.

Reply:
Good catch, thanks! 👍